### PR TITLE
feat(db): metadata and annotation boolean values keep their type

### LINKED ISSUE #13880: [Feature]: add bool type to metadata and annotation table values
### Describe the feature

Currently we have a int and a string type for the value in metadata and annotation. If we want to store booleans they are converted to ints but never returned as a boolean.

If we add a value_type boolean we can save/return booleans true/false without the need of converting them ourselves.

### Is your feature request related to a problem?

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

https://github.com/Elgg/Elgg/issues/6432

### PR DESCRIPTION
**Pull request by jdalsem**
fixes #13880

### CODE REVIEW

File: engine/lib/constants.php [hunk 1]
```diff
@@ -128,6 +128,7 @@
 define('ELGG_VALUE_GUID', 'guid');
 define('ELGG_VALUE_ID', 'id');
 define('ELGG_VALUE_TIMESTAMP', 'timestamp');
+define('ELGG_VALUE_BOOL', 'bool');
```
Comment (jdalsem) [inline engine/lib/constants.php:131]: hmm... not used... should we keep it?

File: engine/classes/Elgg/Database/MetadataTable.php [hunk 2]
```diff
@@ -371,7 +371,9 @@ public function getRowsForGuids(array $guids) {
 			->addOrderBy('time_created', 'asc')
 			->addOrderBy('id', 'asc');
 		
-		return $this->db->getData($qb);
+		return $this->db->getData($qb, function ($row) {
```
Comment (jeabakker) [inline engine/classes/Elgg/Database/MetadataTable.php:374]: you need to update the function docs return type

File: engine/classes/ElggExtender.php [hunk 3]
```diff
@@ -229,11 +228,15 @@ public function getURL() {
...
-		if ($value_type === 'integer' || $value_type === 'text') {
+	public static function detectValueType($value, string $value_type = ''): string {
+		if (in_array($value_type, ['integer', 'text', 'bool'])) {
 			return $value_type;
 		}
 
-		return is_int($value) ? 'integer' : 'text';
+		if (is_int($value)) return 'integer';
+
+		if (is_bool($value)) return 'bool';
```
Comment (jeabakker) [inline engine/classes/ElggExtender.php:238]: please use `if() {} elseif() {}`

File: engine/lib/constants.php (hunk 1 above)
Comment (jeabakker) [reply engine/lib/constants.php:131]: you use this in the `QueryBuilder` :)

File: engine/schema/migrations/20220125084422_add_boolean_value_types.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+     * Remember to call "create()" or "update()" and NOT "save()" when working
+     * with the Table class.
+     */
+    public function change(): void {
+    	
+    	$metadata = $this->table('metadata');
+    	$metadata->changeColumn('value_type', 'enum', ['values' => ['integer', 'text', 'bool']])->update();
+
+    	$annotations = $this->table('annotations');
+    	$annotations->changeColumn('value_type', 'enum', ['values' => ['integer', 'text', 'bool']])->update();
```
Comment (jeabakker) [inline engine/schema/migrations/20220125084422_add_boolean_value_types.php:24]: indentation

**Review by jeabakker — CHANGES_REQUESTED** [review]

**Review by jeabakker — APPROVED** [review]

**Pull request merged**